src/pmdas/bpf/modules/execsnoop.c

```diff
@@ -0,0 +1,395 @@
+        env.fails = atoi(val);
+    if ((val = pmIniFileLookup(cfg, module_name, "command")))
+        env.name = val;
+    if ((val = pmIniFileLookup(cfg, module_name, "line")))
+        env.line = val;
+
+    obj = execsnoop_bpf__open();
+    pmNotifyErr(LOG_INFO, "booting: %s", obj->skeleton->name);
+    if (!obj) {
+        pmNotifyErr(LOG_ERR, "failed to open BPF object");
+        return 1;
+    }
+
+    /* initialize global data (filtering options) */
+    obj->rodata->ignore_failed = !env.fails;
+    obj->rodata->targ_uid = env.uid;
+    obj->rodata->max_args = env.max_args;
+
+    /* initialize global data (filtering options) */
+    obj->rodata->ignore_failed = !env.fails;
+    obj->rodata->targ_uid = env.uid;
+    obj->rodata->max_args = env.max_args;
+
+    err = execsnoop_bpf__load(obj);
+    if (err) {
+        pmNotifyErr(LOG_ERR, "failed to load BPF object: %d", err);
+        return err != 0;
+    }
+
+    err = execsnoop_bpf__attach(obj);
+    if (err) {
+        pmNotifyErr(LOG_ERR, "failed to attach BPF programs");
+        return err != 0;
+    }
+
+    /* setup event callbacks */
+    pb = perf_buffer__new(bpf_map__fd(obj->maps.events), PERF_BUFFER_PAGES,
+            handle_event, handle_lost_events, NULL, NULL);
+    if (!pb) {
+        err = -errno;
+        pmNotifyErr(LOG_ERR, "failed to open perf buffer: %d", err);
+        return err != 0;
+    }
+
+    /* internal/external instance ids */
+    fill_instids(env.process_count, &execsnoop_instances);
+
+    /* Initialize the tail queue. */
+    TAILQ_INIT(&head);
+
+    return err != 0;
+}
+
+void execsnoop_shutdown()
+{
+    free(execsnoop_instances);
+    perf_buffer__free(pb);
+    execsnoop_bpf__destroy(obj);
+    /* Free the entire cache queue. */
+    while ((itemp = TAILQ_FIRST(&head))) {
+        TAILQ_REMOVE(&head, itemp, entries);
+        free(itemp);
+    }
+}
+
+void execsnoop_refresh(unsigned int item)
+{
+    perf_buffer__poll(pb, PERF_POLL_TIMEOUT_MS);
+    itemp = TAILQ_LAST(&head, tailhead);
+}
+
+int execsnoop_fetch_to_atom(unsigned int item, unsigned int inst, pmAtomValue *atom)
+{
+    if (inst == PM_IN_NULL) {
+        return PM_ERR_INST;
+    }
+
+    /* cache buffer is empty */
+    if (itemp == NULL)
+        return PMDA_FETCH_NOVALUES;
+
+    /* bpf.execsnoop.comm */
+    if (item == COMM) {
+        atom->cp = itemp->event.comm;
+    }
+    /* bpf.execsnoop.pid */
+    if (item == PID) {
+        atom->ul = itemp->event.pid;
+    }
+    /* bpf.execsnoop.ppid */
+    if (item == PPID) {
+        atom->ul = itemp->event.ppid;
+    }
+    /* bpf.execsnoop.ret */
+    if (item == RET) {
+        atom->ul = itemp->event.retval;
+    }
+    /* bpf.execsnoop.args */
+    if (item == ARGS) {
+        atom->cp = itemp->event.args;
+    }
+    /* bpf.execsnoop.uid */
+    if (item == UID) {
+        atom->ul = itemp->event.uid;
+    }
+
+    itemp = TAILQ_PREV(itemp, tailhead, entries);
```

Comment:
I think this will never happen because I have a fixed size buffer with max = `env.process_count` elements (this is max number of events I need) and always save the count of elements of this buffer into a global variable `queuelength` (this is the actual number of events.)
if `queuelength` > `env.process_count`, the oldest element will drop & free from the list.
and I start overwriting this buffer all the time with the latest events, 
This seems like a circular buffer,
and I reset the head whole point to the first element of that buffer every `_refresh`.
and I loop over this buffer while through `_fetch_to_atom`

Also, I did a very simple a draft before starting this If you would like to check the logic behind that:
https://github.com/smalinux/tailq-example
`$ clone & make & ./tailq_ex`

Reply:
I was more concerned that this assumes `fetch_to_atom` is called in a specific order. I'm not sure if that is guaranteed to hold in all scenarios as part of the API contract. As per the contract `inst` argument should be used to look up the correct instance in the tailq for each fetch:

```
For each instance listed in the profile (see [pmdaprofile](https://linux.die.net/man/3/pmdaprofile)(3)) of each metric listed in pmidlist, the pmdaFetchCallBack method is called to fill the pmAtomValue structure identified by avp with a value for a specific metric-instance pair identified by the metric descriptor mdesc and the instance inst. This value is then copied into the pmResult structure.
```
https://linux.die.net/man/3/pmdasetfetchcallback

Reply:
From memory I don't think there's a guarantee of any specific inst ordering for the fetch routine.  It'd be worth checking the fetch code below src/libpcp_pmda/src though.

Another option may be to provide a custom fetch routine instead of pmdaFetch if that helps, then you could take control over how the instances are traversed - an example of a PMDA doing this is src/pmdas/sample/src/sample.c